share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -240,9 +249,76 @@ public void startup(
 
         log.info("Starting up.");
         numPartitions = shareGroupTopicPartitionCount.getAsInt();
+        setupRecordPruning();
         log.info("Startup complete.");
     }
 
+    // visibility for tests
+    void setupRecordPruning() {
+        timer.add(new TimerTask(config.shareCoordinatorTopicPruneIntervalMs()) {
+            @Override
+            public void run() {
+                List<CompletableFuture<Void>> futures = new ArrayList<>();
+                runtime.activeTopicPartitions().forEach(tp -> {
+                    futures.add(performRecordPruning(tp));
+                });
+
+                // None of the futures will complete exceptionally.
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{}))
+                    .thenRun(() -> {
+                        // perpetual recursion
+                        setupRecordPruning();
+                    });
```

Comment:
You could omit them here too. I also wonder whether we should re-reschedule in case of failure too. `thenRun` would only reschedule if the future succeeds. It would make it more robust to unexpected failures.